samples/bluetooth/iso_broadcast/src/main.c

```diff
@@ -197,22 +215,116 @@ int main(void)
 			net_buf_reserve(buf, BT_ISO_CHAN_SEND_RESERVE);
 			sys_put_le32(iso_send_count, iso_data);
 			net_buf_add_mem(buf, iso_data, sizeof(iso_data));
-			ret = bt_iso_chan_send(&bis_iso_chan[chan], buf, seq_num);
+
+			/* Read Tx Sync interface returns error if no ISO data was given previously.
+			 */
+			ret = bt_iso_chan_get_tx_sync(&bis_iso_chan[chan], &tx_info);
+			if (ret) {
+				printk("Failed to get tx sync before send (%d)\n", ret);
+
+				/* Use sequence number 0, to synchronize current stream with the
+				 * current payload counter in the ISO group of streams.
+				 */
+				is_first_iso_data_send = true;
+				seq_num[chan] = 0U;
+			}
+
+			/* Send ISO data with timestamp, use sequence number 0 for first ISO data on
+			 * any stream.
+			 */
+			if (!is_first_iso_data_send) {
+				ret = bt_iso_chan_send_ts(&bis_iso_chan[chan], buf, seq_num[chan],
+							  ts);
+			} else {
+				ret = bt_iso_chan_send(&bis_iso_chan[chan], buf, 0U);
+			}
+
 			if (ret < 0) {
 				printk("Unable to broadcast data on channel %u"
 				       " : %d", chan, ret);
 				net_buf_unref(buf);
 				return 0;
 			}
 
+			/* Get Tx Sync, to verify that all streams have their ISO data at the same
+			 * timestamp.
+			 */
+			ret = bt_iso_chan_get_tx_sync(&bis_iso_chan[chan], &tx_info);
+			if (ret) {
+				printk("Failed to get tx sync after send (%d)\n", ret);
+				return 0;
+			}
+
+			/* Timestamp delta with the Controller */
+			if (!delta) {
+				delta = tx_info.ts - ts;
+			} else {
+				int64_t drift = ((delta - tx_info.ts + ts) / BIG_SDU_INTERVAL_US) /
+						1000000U;
+
+				/* NOTE: As the `ts` is just a increase in software in this sample,
+				 *       we will not see any drift.
+				 */
+				if (drift) {
+					printk("Drift: %d ppm.\n", drift);
+				}
+
+				/* NOTE: Change in delta (drift) for a duration of SDU intervals can
+				 *       be used to synchronize the clock with the Controller.
+				 */
+				static uint8_t countdown = 100U;
+				static int64_t drift_us;
+
+				drift_us += delta - tx_info.ts + ts;
+
+				if (!drift_comp_us && !countdown--) {
+					countdown = 100U;
+					drift_comp_us = drift_us / (countdown + 1U);
+					drift_us = 0U;
+
+					printk("Drift Compensate %d us\n", drift_comp_us);
+				}
+			}
+
+			printk("Tx chan %u: seq num %u -> %u ts %u -> %u offset %u\n", chan,
+			       seq_num[chan], tx_info.seq_num, ts, tx_info.ts, tx_info.offset);
+
+			/* Permit a reasonable clock jitter (at least Bluetooth active clock
+			 * jitter), in the verification of timestamp.
+			 */
+			uint32_t jitter_us = BIG_SDU_INTERVAL_US;
```

Comment:
Allowing a jitter of an entire SDU interval seems a bit too much :)

Reply:
The `k_sleep()` (based on kernel system clock) is very flakky, and has large jitters!